# 📘 Course Two: Get Started with Python  
## PACE Strategy Document

## Course 2 tasks:

-Import data

-Create a dataframe 

-Inspect data 

-Share an executive summary with the Waze data team 

*Note*: The story, all names, characters, and incidents portrayed in this project are fictitious. No identification with actual persons (living or deceased) is intended or should be inferred. And, the data shared in this project has been created for pedagogical purposes. 

**Key Takeaways**
In Course 2, Get Started with Python, you were introduced to some basics of the Python programming language. You explored syntax, loops, strings, lists, dictionaries, object-oriented programming, and explored how data professionals use code on the job. 

**Course 2 skills:**

-Code with Python

-Use comments to enhance code readability

-Work within a Jupyter Notebook

-Share insights and ideas with stakeholders
    
**Course 2 end-of-course project deliverables:** 

_Build a dataframe

_Create an executive summary

_The end-of-course portfolio projects are designed for you to apply your data analytical skills within a workplace scenario. No matter which scenario you work with, you will practice your ability to discuss data analytic topics with coworkers, internal team members, and external clients.

_As a reminder, you are required to complete one project for each course. To gain additional practice, or to add more samples to your portfolio, you may complete as many of the scenarios as you wish.

### **Main task: Data reviewing**

**Course 2 Waze project lab:**

Import, inspect, and organize data 

### 🧭 PLAN Stage

**How can you best prepare to understand and organize the provided information?**  
To prepare, I reviewed the project proposal, referenced communications from stakeholders (e.g., May Santner and Chidi Ga), and downloaded the Waze synthetic churn dataset. I ensured that Python libraries like `pandas`and `numpy`were installed. I set up my Jupyter environment, examined the dataset structure, reviewed column names, and loaded the data using its full file path.

**What follow-along and self-review codebooks will help you perform this work?**  
I referenced Python notebooks from Course 1 and 2, and reviewed similar churn prediction projects on Kaggle and GitHub.

**What are some additional activities a resourceful learner would perform before starting to code?**  
- Reviewing documentation for relevant data wrangling functions  

## **Task 1. Understand the situation**
Prepare by reading in the data, viewing the data dictionary, and exploring the dataset to identify key variables for the stakeholder.

## **Task 2a.Imports and data loading**
Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

import pandas as pd

import numpy as np

In [7]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

In [8]:
# Load the dataset
df = pd.read_csv(r'D:\5B. Google_Advanced_data_analysis\training_project_data\waze_dataset.csv')

In [9]:
df.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

In [10]:
# Datafram shape
print("\nDataset shape:")
df.shape


Dataset shape:


(14999, 13)

In [11]:
# Summary statistics for numeric data
print("\nNumeric Summary:\n")
df.describe()


Numeric Summary:



ID      sessions        drives  total_sessions  \
count  14999.000000  14999.000000  14999.000000    14999.000000   
mean    7499.000000     80.633776     67.281152      189.964447   
std     4329.982679     80.699065     65.913872      136.405128   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.661156   
50%     7499.000000     56.000000     48.000000      159.568115   
75%    11248.500000    112.000000     93.000000      254.192341   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14999.000000            14999.000000   
mean               1749.837789              121.605974   
std                1008.513876              148.121544   
min                   4.000000                0.000000   
25%                 878.000000                9.000000   
50%                1741.000000               71.000000   
75%                2623.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14999.000000      14999.000000             14999.000000   
mean                29.672512       4039.340921              1860.976012   
std                 45.394651       2502.149334              1446.702288   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2212.600607               835.996260   
50%                  9.000000       3493.858085              1478.249859   
75%                 43.000000       5289.861262              2464.362632   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14999.000000  14999.000000  
mean       15.537102     12.179879  
std         9.004655      7.824036  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

In [12]:
# Display dataset information
print("\nDataset Information:")
print(df.info())


Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ 

In [13]:
#Finding items with missing values if any
print("\nMissing values per column:")
df.isnull().sum()


Missing values per column:


ID                           0
label                      700
sessions                     0
drives                       0
total_sessions               0
n_days_after_onboarding      0
total_navigations_fav1       0
total_navigations_fav2       0
driven_km_drives             0
duration_minutes_drives      0
activity_days                0
driving_days                 0
device                       0
dtype: int64

_The attribute `label` have 700 missing values.

_Threre are 13 attributes, almost all of them contains 14999 variables aside `label`. Data types include float(3), int(8),and object.

In [15]:
# Group columns by data types
column_groups = {
    'int64': df.select_dtypes(include='int64').columns.tolist(),
    'float64': df.select_dtypes(include='float64').columns.tolist(),
    'object': df.select_dtypes(include='object').columns.tolist()
}

# Print grouped column names
for dtype, columns in column_groups.items():
    print(f"\nColumns with data type '{dtype}':")
    for col in columns:
        print(f"  - {col}")


Columns with data type 'int64':
  - ID
  - sessions
  - drives
  - n_days_after_onboarding
  - total_navigations_fav1
  - total_navigations_fav2
  - activity_days
  - driving_days

Columns with data type 'float64':
  - total_sessions
  - driven_km_drives
  - duration_minutes_drives

Columns with data type 'object':
  - label
  - device


## **Task 2c. Null values and summary statistics**
Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question**: Is there a discernible difference between the two populations?

In [17]:
# Isolate rows with null values
null_df = df[df['label'].isnull()]
# Display summary stats of rows with null values
null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [18]:
# Isolate rows without null values
not_null_values = df[df['label'].notnull()]
# Display summary stats of rows without null values
not_null_values.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

There is no remarkable difference between with and without null values tables, perharps because there is sufficient amount of samples. 

## **Task 2d. Null values - device counts**
Next, check the two populations with respect to the device variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [21]:
# Get count of null values by device
null_df['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

There are 447 iPhone users have null values and 253 Android users have null values.

Now, of the rows with null values, calculate the percentage with each device—Android and iPhone. You can do this directly with the value_counts() function.

In [24]:
# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize = True)*100

device
iPhone     63.857143
Android    36.142857
Name: proportion, dtype: float64

How does this compare to the device ratio in the full dataset?

In [26]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)*100

device
iPhone     64.484299
Android    35.515701
Name: proportion, dtype: float64

The percentages with each device—Android and iPhone are similar to the devices ratio in the full dataset

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [29]:
# Calculate counts of churned vs. retained
print(df['label'].value_counts())
print()
print(df['label'].value_counts(normalize=True)*100)

label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    82.264494
churned     17.735506
Name: proportion, dtype: float64


This dataset contains 82% retained users and 18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the driven_km_drives column is 21,183 km. That's more than half the circumference of the earth!

In [32]:
# Calculate median values of all columns for churned and retained user
print("\nMedian values grouped by label:")
df.groupby('label').median(numeric_only=True)


Median values grouped by label:


ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

**Average Insights from Last Month's Data:**

Churned users drove, on average, 3 km more than retained users over the past month on an occurrence.

On average, churned users had used the app for 500 fewer days than retained users, calculated from the date of signup.

Retained users had a 150-minute shorter total driving duration during the month compared to churned users.

However, the number of active days for retained users was nearly double that of churned users.

The number of driving days for retained users was more than half that of churned users.
                                                                                                                            

Calculate the median kilometers per drive in the last month for both retained and churned users.

In [35]:
# Select only numeric columns to avoid TypeError
numeric_cols = df.select_dtypes(include='number').columns

print("Median kilometers per drive in the last month for both retained and churned users")

# Group by 'label' and calculate the median for numeric columns only
median_by_label = df.groupby('label')[numeric_cols].median()

# Divide the median distance by the median number of drives
result = median_by_label['driven_km_drives'] / median_by_label['drives']
print(result)

Median kilometers per drive in the last month for both retained and churned users
label
churned     73.053113
retained    73.716694
dtype: float64


The median kilometers per drive in the last month for both retained and churned users are both approximately about 73 km/drive.


How many kilometers per driving day was this?

In [38]:
# Divide the median distance by median number of driving days
median_by_label['driven_km_drives']/median_by_label['driving_days']

label
churned     608.775944
retained    247.477472
dtype: float64

Now, calculate the median number of drives per driving day for each group.

In [40]:
# Divide the median number of drives by median number of driving days
median_by_label['drives']/median_by_label['driving_days']

label
churned     8.333333
retained    3.357143
dtype: float64

According to last month’s data, churned users had a median daily driving distance of 608 kilometers on the days they drove, which is 247% of the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.​

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data—and in particular the sample of churned users—contains a high proportion of long-haul truckers.​

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.​

Begin by getting the overall counts of each device type for each group, churned and retained.

In [43]:
# For each label, calculate the number of Android users and iPhone users
df.groupby(['label','device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [45]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True) * 100

label     device 
churned   iPhone     64.865931
          Android    35.134069
retained  iPhone     64.439344
          Android    35.560656
Name: proportion, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

### 📊 ANALYZE Stage

**Will the available information be sufficient to achieve the goal based on your intuition and the analysis of the variables?**  
Yes, the dataset is informative for churn prediction.
    
**How would you build summary dataframe statistics and assess the min and max range of the data?**  
I used `df.describe()` for summary statistics on numeric columns, `as well as assess the min and max range of the data for this section only.

**Do the averages of any of the data variables look unusual? Can you describe the interval data?**  
Yes, some averages—like trip count or distance—are skewed by outliers. Interval variables such as number of driving days and kilometers per drive have wide ranges. 

### 🛠 CONSTRUCT Stage  
*Note: This stage does not apply directly to the current workflow, as the focus is on exploratory analysis rather than building a final product or visualization. It may apply in future phases of the project.*

### 🚀 EXECUTE Stage

**Given your current knowledge of the data, what would you initially recommend to your manager to investigate further prior to performing exploratory data analysis?**  
I would focus on user engagement patterns like trip frequency, session counts, and recency of use. Additionally, I’d review data quality for users who appear inactive yet are not labeled as churned, and seek clarity on the churn labeling logic.

**What data initially presents as containing anomalies?**  
Columns with data type 'int64':
  - drives
  - n_days_after_onboarding
  - activity_days
  - driving_days
  - driven_km_drives
  - duration_minutes_drives

**What additional types of data could strengthen this dataset?**  
- Marketing engagement (e.g., email opens, push responses)  
- Regional location data for user segmentation  
- App version or update history  
- User satisfaction ratings or survey results  
- Competitor app usage trends

## **Task 3. Conclusion**
**Questions:**

Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?

What is a benefit of using the median value of a sample instead of the mean?

Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?

What percentage of the users in the dataset were Android users and what percentage were iPhone users?

What were some distinguishing characteristics of users who churned vs. users who were retained?

Was there an appreciable difference in churn rate between iPhone users vs. Android users?

(1) There are 700 value missing in label column. No pattern found to the missing values.

(2) Mean is affected by outliers values but median always shows the middle value regardless of outlying values.

(3) Yes, I would like to explore how the outliers affect the data.
    Also, I want to know the duration that they use for data collecting.

(4) Android users : 36% and iPhone users : 64%

(5) Users who churned drove farther and longer than retained users, nontheless, their active days and driving days are less than that of the retained users.

(6) There is no appreciable difference in churn rate between iPhone users and Android users.

In [2]:
# Step 1: In your local Jupyter Notebook, export your file
# This code helps you download a file from your Jupyter environment to your local machine
from IPython.display import display, Javascript
import os

def download_file(file_path):
    """Generate a link to download a file from Jupyter."""
    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' not found.")
        return
    
    file_name = os.path.basename(file_path)
    
    # Display file info
    file_size = os.path.getsize(file_path) / (1024 * 1024)  # Convert to MB
    print(f"File: {file_name} ({file_size:.2f} MB)")
    
    # Create download link
    from IPython.display import HTML
    html = f"""
    <a href="files/{file_name}" download="{file_name}">
        Download {file_name}
    </a>
    """
    display(HTML(html))